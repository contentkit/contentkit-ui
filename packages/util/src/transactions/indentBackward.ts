import { leftIndent } from 'left-indent'
import { ChangeType } from '../fixtures'
import setEditorStateBlockMap from './setEditorStateBlockMap'
import invariant from 'fbjs/lib/invariant'
import { EditorState, ContentBlock } from 'draft-js'

const indentBackward = (editorState: EditorState, tabSize = 2): EditorState => {
  invariant(!isNaN(tabSize), 'tabSize is not a number')
  const contentState = editorState.getCurrentContent()
  const selectionState = editorState.getSelection()
  const focusKey = selectionState.getFocusKey()
  const blockMap = contentState.getBlockMap()
  const focusBlock = blockMap.get(focusKey)
  const text = focusBlock.getText()
  const offset = selectionState.getFocusOffset()
  const textAfter = text.slice(offset, text.length)
  const textBefore = leftIndent(text.slice(0, offset), 'backward')
  const chars = focusBlock.getCharacterList()
  const focusOffset = textBefore.length
  const blockProperties = {
    text: `${textBefore}${textAfter}`,
    characterList: chars.slice(0, focusOffset).concat(chars.slice(offset, text.length))
  }
  const nextBlockMap = blockMap.set(focusKey, (focusBlock.merge(blockProperties) as ContentBlock))
  return setEditorStateBlockMap(editorState, nextBlockMap, focusKey, textBefore.length)
}

export default indentBackward
