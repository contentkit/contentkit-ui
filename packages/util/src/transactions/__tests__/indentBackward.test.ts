import indentBackward from '../indentBackward'
import { EditorState, SelectionState, ContentState, ContentBlock, CharacterMetadata } from 'draft-js'
import { List, Repeat } from 'immutable'
import { pad } from 'left-indent'

const getEditorState = (blockArray, focusKey, focusOffset) => {
  const editorState = EditorState.createWithContent(
    ContentState.createFromBlockArray(blockArray)
  )
  return EditorState.acceptSelection(editorState,
    new SelectionState({
      anchorKey: focusKey,
      anchorOffset: focusOffset,
      focusKey: focusKey,
      focusOffset: focusOffset,
      isBackward: false
    })
  )
}

const samples = [
  `${pad(2)}bar `,
  `${pad(4)}foo `,

]

it('indentBackward should correctly handle', () => {
  const text = `${pad(2)}bar `
  const blocksArray = [
    new ContentBlock({
      key: 'A',
      text,
      characterList: List(Repeat(new CharacterMetadata(), text.length))
    })
  ]

  const editorState = getEditorState(blocksArray, 'A', 2)
  const block = indentBackward(editorState).getCurrentContent().getBlockMap().get('A')
  expect(block).toMatchSnapshot()
})

it('indentBackward should remove 1 space', () => {
  const text = `${pad(3)}foo `
  const blocksArray = [
    new ContentBlock({
      key: 'A',
      text: text,
      characterList: List(Repeat(new CharacterMetadata(), text.length))
    })
  ]

  const editorState = getEditorState(blocksArray, 'A', 3)
  const block = indentBackward(editorState).getCurrentContent().getBlockMap().get('A')
  expect(block.getText()).toMatchSnapshot()
  expect(block.getText()).toMatch(`${pad(2)}foo `)
})

it('indentBackward should remove 2 spaces', () => {
  const text = `${pad(4)}foo `
  const blocksArray = [
    new ContentBlock({
      key: 'A',
      text: text,
      characterList: List(Repeat(new CharacterMetadata(), text.length))
    })
  ]

  const editorState = getEditorState(blocksArray, 'A', 4)
  const block = indentBackward(editorState).getCurrentContent().getBlockMap().get('A')
  expect(block.getText()).toMatchSnapshot()
  expect(block.getText()).toMatch(`${pad(2)}foo `)
})

it('indentBackward should remove 2 spaces', () => {
  const text = pad(2)
  const blocksArray = [
    new ContentBlock({
      key: 'A',
      text: text,
      characterList: List(Repeat(new CharacterMetadata(), text.length))
    })
  ]

  const editorState = getEditorState(blocksArray, 'A', 2)
  const block = indentBackward(editorState).getCurrentContent().getBlockMap().get('A')
  expect(block.getText()).toMatchSnapshot()
  expect(block.getText()).toMatch('')
})



it('indentBackward should remove 2 spaces', () => {
  const text = pad(2)
  const blocksArray = [
    new ContentBlock({
      key: 'A',
      text: text,
      characterList: List(Repeat(new CharacterMetadata(), text.length))
    })
  ]

  const editorState = getEditorState(blocksArray, 'A', 2)
  const block = indentBackward(editorState).getCurrentContent().getBlockMap().get('A')
  expect(block.getText()).toMatchSnapshot()
  expect(block.getText()).toMatch('')
})


