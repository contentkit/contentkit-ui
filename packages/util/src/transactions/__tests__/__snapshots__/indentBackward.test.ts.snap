// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indentBackward should correctly handle 1`] = `
Immutable.Record {
  "key": "A",
  "type": "unstyled",
  "text": "bar ",
  "characterList": Immutable.List [
    Immutable.Record {
      "style": Immutable.OrderedSet [],
      "entity": null,
    },
    Immutable.Record {
      "style": Immutable.OrderedSet [],
      "entity": null,
    },
    Immutable.Record {
      "style": Immutable.OrderedSet [],
      "entity": null,
    },
    Immutable.Record {
      "style": Immutable.OrderedSet [],
      "entity": null,
    },
  ],
  "depth": 0,
  "data": Immutable.Map {},
}
`;

exports[`indentBackward should remove 1 space 1`] = `"  foo "`;

exports[`indentBackward should remove 2 spaces 1`] = `"  foo "`;

exports[`indentBackward should remove 2 spaces 2`] = `""`;

exports[`indentBackward should remove 2 spaces 3`] = `""`;
